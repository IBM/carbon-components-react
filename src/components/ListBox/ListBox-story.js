import React from 'react';
import { storiesOf, action } from '@storybook/react';
import Combobox from '../Combobox';
import Dropdown from '../DropdownV2';
import MultiSelect from '../MultiSelect';
import FilterableMultiSelect from '../FilterableMultiSelect';

const items = [
  {
    id: 'option-1',
    text: 'Option 1',
  },
  {
    id: 'option-2',
    text: 'Option 2',
  },
  {
    id: 'option-3',
    text: 'Option 3',
  },
  {
    id: 'option-4',
    text: 'Option 4',
  },
];

storiesOf('ListBox', module).addWithInfo(
  'default',
  `
    MultiSelect and Combobox
  `,
  () => (
    <div>
      <div style={{ display: 'flex', alignItems: 'center' }}>
        <Dropdown
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange - Dropdown')}
        />
        <div style={{ marginRight: '2rem' }} />
        <Dropdown
          type="inline"
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange - Dropdown')}
        />
        <div style={{ marginRight: '2rem' }} />
        <Dropdown
          type="inline"
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange - Dropdown')}
          disabled
        />
      </div>
      <div style={{ marginBottom: '2rem' }} />
      <div style={{ display: 'flex', alignItems: 'center' }}>
        <MultiSelect
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange - MultiSelect')}
        />
        <div style={{ marginRight: '2rem' }} />
        <MultiSelect
          type="inline"
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange - Inline MultiSelect')}
        />
        <div style={{ marginRight: '2rem' }} />
        <MultiSelect
          type="inline"
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange - Inline MultiSelect')}
          disabled
        />
      </div>
      <div style={{ marginBottom: '2rem' }} />
      <div style={{ display: 'flex', alignItems: 'center' }}>
        <Combobox
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange - Combobox')}
          placeholder="Filter..."
        />
        <div style={{ marginRight: '2rem' }} />
        <Combobox
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange - Combobox')}
          placeholder="Filter..."
          disabled
        />
      </div>
      <div style={{ marginBottom: '2rem' }} />
      <div style={{ display: 'flex', alignItems: 'center' }}>
        <FilterableMultiSelect
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange - Combobox')}
          placeholder="Filter..."
        />
        <div style={{ marginRight: '2rem' }} />
        <FilterableMultiSelect
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange - Combobox')}
          placeholder="Filter..."
          disabled
        />
      </div>
    </div>
  )
);
