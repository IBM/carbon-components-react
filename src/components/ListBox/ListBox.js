import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

const ListBox = ({
  type = 'default',
  innerRef,
  children,
  className: containerClassName,
  disabled,
  selectionType,
}) => {
  const className = cx(containerClassName, {
    'bx--list-box': true,
    'bx--list-box--inline': type === 'inline',
    'bx--list-box--disabled': disabled,
    'bx--list-box--multiselect': selectionType === 'multiselect',
  });
  return (
    <div className={className} ref={innerRef}>
      {children}
    </div>
  );
};

ListBox.propTypes = {
  type: PropTypes.oneOf(['default', 'inline']),
  innerRef: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
};

export default ListBox;
