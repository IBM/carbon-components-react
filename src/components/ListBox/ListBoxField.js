import React from 'react';
import PropTypes from 'prop-types';

const ListBoxField = ({ children, ...rest }) => (
  <div role="button" className="bx--list-box__field" tabIndex="0" {...rest}>
    {children}
  </div>
);

ListBoxField.propTypes = {
  children: PropTypes.node,
};

export default ListBoxField;
