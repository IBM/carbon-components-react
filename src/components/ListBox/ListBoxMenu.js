import React from 'react';
import PropTypes from 'prop-types';

const ListBoxMenu = ({ children, ...rest }) => {
  return (
    <div className="bx--list-box__menu" {...rest}>
      {children}
    </div>
  );
};

ListBoxMenu.propTypes = {
  children: PropTypes.node,
};

export default ListBoxMenu;
