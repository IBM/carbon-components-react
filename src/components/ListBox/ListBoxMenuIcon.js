import cx from 'classnames';
import React from 'react';
import PropTypes from 'prop-types';
import Icon from '../Icon';

const defaultTranslations = {
  'close.menu': 'Close menu',
  'open.menu': 'Open menu',
};

const ListBoxMenuIcon = ({ isOpen, translateWithId: t }) => {
  const className = cx({
    'bx--list-box__menu-icon': true,
    'bx--list-box__menu-icon--open': isOpen,
  });
  const description = isOpen ? t('close.menu') : t('open.menu');
  return (
    <div className={className}>
      <Icon name="caret--down" description={description} />
    </div>
  );
};

ListBoxMenuIcon.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  translateWithId: PropTypes.func.isRequired,
};

ListBoxMenuIcon.defaultProps = {
  translateWithId: id => defaultTranslations[id],
};

export default ListBoxMenuIcon;
