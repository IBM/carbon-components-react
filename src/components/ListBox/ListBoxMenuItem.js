import cx from 'classnames';
import React from 'react';
import PropTypes from 'prop-types';

const ListBoxMenuItem = ({ children, isActive, isHighlighted, ...rest }) => {
  const className = cx({
    'bx--list-box__menu-item': true,
    'bx--list-box__menu-item--active': isActive,
    'bx--list-box__menu-item--highlighted': isHighlighted,
  });
  return (
    <div className={className} {...rest}>
      {children}
    </div>
  );
};

ListBoxMenuItem.propTypes = {
  children: PropTypes.node,
  isActive: PropTypes.bool.isRequired,
  isHighlighted: PropTypes.bool.isRequired,
};

ListBoxMenuItem.defaultProps = {
  isActive: false,
  isHighlighted: false,
};

export default ListBoxMenuItem;
