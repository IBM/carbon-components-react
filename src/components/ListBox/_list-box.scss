@import 'carbon-components/scss/globals/scss/colors';
@import 'carbon-components/scss/globals/scss/import-once';
@import 'carbon-components/scss/globals/scss/css--helpers';
@import 'carbon-components/scss/globals/scss/vars';
@import 'carbon-components/scss/globals/scss/layout';

$list-box-width: rem(300px);
$list-box-height: rem(40px);
$list-box-inline-height: rem(32px);

@include exports('list-box') {
  .bx--list-box {
    position: relative;
    width: $list-box-width;
    height: $list-box-height;
    background-color: rgba($field-01, 0.1);
  }

  .bx--list-box--disabled {
    opacity: 0.5;
  }

  .bx--list-box input[role='combobox'] {
    background-color: inherit;
  }

  .bx--list-box input[role='combobox'] {
    outline-offset: 0;
  }

  .bx--list-box--disabled input[role='combobox'] {
    opacity: 1;
  }

  .bx--list-box__menu {
    @include box-shadow();

    position: absolute;
    left: 0;
    right: 0;
    top: $list-box-height;
    background-color: $ui-01;
    max-height: rem(120px);
    overflow-y: auto;
    z-index: z('dropdown');
  }

  .bx--list-box__menu-item {
    display: flex;
    align-items: center;
    font-size: rem(14px);
    height: rem(40px);
    padding: 0 1rem;
    cursor: pointer;
    user-select: none;
  }

  .bx--list-box__menu-item:hover,
  .bx--list-box__menu-item--highlighted {
    background-color: $brand-01;
    color: white;
  }

  .bx--list-box__menu-icon {
    display: inline-block;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    height: 100%;
    padding: 0 1rem;
    transition: transform 200ms $bx--standard-easing;
    cursor: pointer;
  }

  .bx--list-box__menu-icon > svg {
    fill: $brand-01;
    height: 100%;
  }

  .bx--list-box__menu-icon--open {
    transform: rotate(180deg);
  }

  .bx--list-box__selection {
    display: inline-block;
    position: absolute;
    top: 0;
    right: rem(26px);
    bottom: 0;
    height: rem(40px);
    padding: 0 1rem;
    cursor: pointer;
    z-index: z('overlay');
    user-select: none;
  }

  .bx--list-box__selection > svg {
    fill: $ui-05;
    height: 100%;
  }

  .bx--list-box__selection:focus {
    outline: 1px solid $brand-01;
  }

  .bx--list-box__selection--multi {
    position: static;
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 0;
    background-color: $brand-01;
    height: rem(18px);
    color: white;
    font-size: 12px;
    font-weight: 200;
    padding: rem(5px);
    margin-right: rem(10px);
    border-radius: rem(13px);
  }

  .bx--list-box__selection--multi > svg {
    fill: white;
    width: 0.5rem;
    height: 0.5rem;
    margin-left: rem(5px);
  }

  .bx--list-box__selection--multi:focus,
  .bx--list-box__selection--multi:hover {
    background-color: $brand-02;
    outline: none;
  }

  .bx--list-box__field {
    @include button-reset;
    position: relative;
    display: flex;
    align-items: center;
    width: 100%;
    height: 100%;
    padding: 0 1rem;
    cursor: pointer;
  }

  .bx--list-box__field:focus,
  .bx--list-box__field[aria-expanded='true'] {
    outline: 1px solid $brand-01;
  }

  .bx--list-box__field[disabled] {
    outline: none;
  }

  .bx--list-box__field[aria-expanded='true'] ~ .bx--list-box__menu {
    box-shadow: 0 1px 0 0 $brand-01, 1px 0 0 0 $brand-01, -1px 0 0 0 $brand-01,
      0 4px 8px 0 rgba(0, 0, 0, 0.1);
  }

  .bx--combo-box > .bx--list-box__field {
    padding: 0;
  }

  .bx--list-box__label {
    color: $text-01;
    font-size: 14px;
    font-weight: 600;
    user-select: none;
  }

  .bx--list-box__menu-item:hover,
  .bx--list-box__menu-item--highlighted {
    background-color: rgba($field-01, 0.1);
    color: $text-01;
  }

  .bx--list-box--inline {
    background-color: inherit;
    width: auto;
    height: $list-box-inline-height;
  }

  .bx--list-box--inline .bx--list-box__label {
    color: $brand-01;
  }

  .bx--list-box--inline .bx--list-box__field {
    padding: 0 0 0 rem(10px);
  }

  .bx--list-box--inline .bx--list-box__menu-icon {
    position: static;
    padding: 0 rem(10px) 0 rem(12px);
  }

  .bx--list-box--inline > .bx--list-box__menu {
    top: $list-box-inline-height;
    width: $list-box-width;
  }

  .bx--list-box--inline
    > .bx--list-box__field[aria-expanded='true']
    ~ .bx--list-box__menu {
    outline: 1px solid $brand-01;
    box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.1);
  }

  .bx--list-box--inline > .bx--list-box__field[aria-expanded='true'] {
    outline: none;
  }

  .bx--dropdown-v2.bx--list-box--inline .bx--list-box__selection {
    position: static;
    padding: 0 rem(10px);
  }
}
