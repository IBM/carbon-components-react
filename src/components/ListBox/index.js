import cx from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import Icon from '../Icon';

export const ListBox = ({
  children,
  className: containerClassName,
  isDisabled,
  innerRef,
  ...rest
}) => {
  const className = cx('bx--list-box', containerClassName, {
    'bx--list-box--disabled': isDisabled,
  });
  return (
    <div className={className} ref={innerRef} {...rest}>
      {children}
    </div>
  );
};

export const ListBoxField = ({ children, ...other }) => (
  <div className="bx--list-box__field" tabIndex="0" {...other}>
    {children}
  </div>
);

export const ListBoxSelection = ({
  clearSelection,
  selectedItem,
  shouldDisplay,
}) => {
  const handleOnKeyDown = event => {
    // When a user hits ENTER, we'll clear the selection
    if (event.keyCode === 13) {
      clearSelection();
    }
  };
  return (
    <div
      className="bx--list-box__selection"
      tabIndex="0"
      onClick={clearSelection}
      onKeyDown={handleOnKeyDown}>
      <Icon name="close" description="Clear current selection" />
    </div>
  );
};

export const ListBoxMenu = ({ children }) => {
  return <div className="bx--list-box__menu">{children}</div>;
};

export const ListBoxMenuItem = ({
  children,
  isActive,
  isHighlighted,
  ...rest
}) => {
  const className = cx({
    'bx--list-box__menu-item': true,
    'bx--list-box__menu-item--active': isActive,
    'bx--list-box__menu-item--highlighted': isHighlighted,
  });
  return (
    <div className={className} {...rest}>
      {children}
    </div>
  );
};

export const ListBoxMenuIcon = ({ isOpen }) => {
  const className = cx({
    'bx--list-box__menu-icon': true,
    'bx--list-box__menu-icon--open': isOpen,
  });
  const description = isOpen ? 'Close menu' : 'Open menu';
  return (
    <div className={className}>
      <Icon name="caret--down" description={description} />
    </div>
  );
};

ListBoxMenuIcon.propTypes = {
  isOpen: PropTypes.bool.isRequired,
};
