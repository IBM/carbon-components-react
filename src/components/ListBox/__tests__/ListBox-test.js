import React from 'react';
import { mount } from 'enzyme';
import ListBox from '../ListBox';

describe('ListBox', () => {
  let mockProps;

  beforeEach(() => {
    mockProps = {
      type: 'default',
      children: <span>Hello</span>,
      className: 'bx--list-box__container',
      isDisabled: false,
      innerRef: jest.fn(),
    };
  });

  it('should render', () => {
    const wrapper = mount(<ListBox {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render an inline class if the type=`inline`', () => {
    const wrapper = mount(<ListBox {...mockProps} type="inline" />);
    expect(wrapper.find('.bx--list-box--inline').length).toBe(1);
  });

  it('should render a disabled class if isDisabled is true', () => {
    const wrapper = mount(<ListBox {...mockProps} isDisabled={true} />);
    expect(wrapper.find('.bx--list-box--disabled').length).toBe(1);
  });

  it('should call the provided `innerRef` function with a ref to the node', () => {
    mount(<ListBox {...mockProps} />);
    expect(mockProps.innerRef).toHaveBeenCalled();
  });

  it('should add the provided `className` to the root node', () => {
    const wrapper = mount(<ListBox {...mockProps} />);
    expect(
      wrapper
        .children()
        .prop('className')
        .includes(mockProps.className)
    ).toBe(true);
  });
});
