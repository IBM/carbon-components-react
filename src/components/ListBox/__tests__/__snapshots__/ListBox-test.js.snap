// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListBox should render 1`] = `
<ListBox
  ariaLabel="Choose an item"
  className="bx--list-box__container"
  disabled={false}
  innerRef={
    [MockFunction] {
      "calls": Array [
        Array [
          <div
            aria-label="Choose an item"
            class="bx--list-box__container bx--list-box"
            role="listbox"
          >
            <div
              class="bx--list-box__field"
              role="button"
              tabindex="0"
            />
          </div>,
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  type="default"
>
  <div
    aria-label="Choose an item"
    className="bx--list-box__container bx--list-box"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="listbox"
    tabIndex={null}
  >
    <ListBoxField>
      <div
        aria-controls={null}
        aria-owns={null}
        className="bx--list-box__field"
        role="button"
        tabIndex="0"
      />
    </ListBoxField>
  </div>
</ListBox>
`;
